### PR TITLE
lib/cpuinfo: Increase the file descriptors limit to handle more CPUs

### PR DESCRIPTION
**Pull request by babumoger**
The pqos tool fails with the following errors on systems with 300 or more CPU cores.
$pqos
NOTE:  Mixed use of MSR and kernel interfaces to manage
       CAT or CMT & MBM may lead to unexpected behavior.
ERROR: Could not open /sys/fs/resctrl directory
ERROR: Failed to stop resctrl events
ERROR: Failed to start all selected OS monitoring events Monitoring start error on core(s) 339, status 1

By default, the file descriptor limit is set to 1024 for a session. pqos monitor uses 3 descriptors for each CPU for perf monitoring. So, it runs out of limit(1024) on systems with 300 or more CPUs.

Fix the issue by detecting the number of CPUs in the system and increasing the descriptor limit using system call getrlimit and setrlimit respectively. Increase the limit to 4 times the number of CPUs to take care of open files limit.

## Description
By default, the file descriptor limit is set to 1024 for a session. pqos monitor uses 3 descriptors for each CPU for perf monitoring. So, it runs out of limit(1024) on systems with 300 or more CPUs.

Fix the issue by detecting the number of CPUs in the system and increasing the descriptor limit using system call getrlimit and setrlimit respectively. Increase the limit to 4 times the number of CPUs to take care of open files limit.

## Affected parts

- [x] library
- [x] pqos utility
- [ ] rdtset utility
- [ ] App QoS
- [ ] other: (please specify)

## Motivation and Context

https://github.com/intel/intel-cmt-cat/issues/261

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by babumoger** [comment]
Please take a look at the code.

**Comment by rkanagar** [comment]
> Please take a look at the code.

Yes, we are reviewing this code. Thanks

**Comment by rkanagar** [comment]
Hi Babu,
   Please implement the following lib/common.c

diff --git a/lib/common.c b/lib/common.c
index 2ea12053..42a59fbf 100644
--- a/lib/common.c
+++ b/lib/common.c
@@ -46,6 +46,11 @@
 #include <sys/resource.h>
 #include <unistd.h>

+/* Maximum required file descriptors per core */
+#define MAX_FD_PER_CORE 4
+/* pqos tool opens some file descriptors while using msr interface */
+#define MAX_PQOS_FD 100
+
 FILE *
 pqos_fopen(const char *name, const char *mode)
 {
@@ -398,16 +403,19 @@ int
 pqos_set_no_files_limit(unsigned long max_core_count)
 {
         struct rlimit files_limit;
+        rlim_t required_fd = (max_core_count * MAX_FD_PER_CORE) + MAX_PQOS_FD;

         if (getrlimit(RLIMIT_NOFILE, &files_limit))
                 return PQOS_RETVAL_ERROR;

-        /* Check if rlim_max can handle the number of CPUs */
-        if (files_limit.rlim_max < max_core_count * 4)
-                return PQOS_RETVAL_ERROR;
+        /* Check Kernel allows to open required file descriptors */
+        if (files_limit.rlim_max < required_fd ||
+            files_limit.rlim_cur < required_fd) {
+                if (files_limit.rlim_max < required_fd)
+                        files_limit.rlim_max = required_fd;
+
+                files_limit.rlim_cur = required_fd;

-        if (files_limit.rlim_cur < max_core_count * 4) {
-                files_limit.rlim_cur = max_core_count * 4;
                 if (setrlimit(RLIMIT_NOFILE, &files_limit))
                         return PQOS_RETVAL_ERROR;
         }

**Comment by babumoger** [comment]
> Hi Babu, Please implement the attached fd_diff.txt [fd_diff.txt](https://github.com/intel/intel-cmt-cat/files/15025725/fd_diff.txt)

Hi Raghavan, I have implemented your changes. Please review. thanks

File: lib/common.c [hunk 1]
```diff
@@ -392,3 +398,27 @@ pqos_read(int fd, void *buf, size_t count)
 
         return count;
 }
+
+int
+pqos_set_no_files_limit(unsigned long max_core_count)
+{
+        struct rlimit files_limit;
+        rlim_t required_fd = (max_core_count * MAX_FD_PER_CORE) + MAX_PQOS_FD;
```
Comment (rkanagar) [inline lib/common.c:406]: Add const in this variable

File: lib/common.c [hunk 2]
```diff
@@ -43,8 +43,14 @@
 #include <stdlib.h>
 #include <string.h>
 #include <sys/mman.h>
+#include <sys/resource.h>
 #include <unistd.h>
 
+/* Maximum required file descriptors per core */
+#define MAX_FD_PER_CORE 4
```
Comment (rkanagar) [inline lib/common.c:50]: The MAX_FD_PER_CORE should be 5.

1 fd to read/write msr
2 fd to obtain IPC value
1 fd to obtain LLC miss value
1 fd to obtain LLC ref value

File: lib/common.c (hunk 1 above)
Comment (babumoger) [reply lib/common.c]: Done

File: lib/common.c (hunk 2 above)
Comment (babumoger) [reply lib/common.c]: Done

File: lib/cpuinfo.c [hunk 3]
```diff
@@ -57,6 +58,7 @@
 #endif
 #ifdef __linux__
 #include <sched.h> /* sched affinity */
+#include <sys/resource.h>
```
Comment (rkanagar) [inline lib/cpuinfo.c:61]: Remove this inclusion. common.c has this file

File: lib/os_cpuinfo.c [hunk 4]
```diff
@@ -43,6 +43,7 @@
 #include <limits.h>
 #include <stdlib.h>
 #include <string.h>
+#include <sys/resource.h>
```
Comment (rkanagar) [inline lib/os_cpuinfo.c:46]: Remove this inclusion. common.c has this file

File: lib/cpuinfo.c (hunk 3 above)
Comment (babumoger) [reply lib/cpuinfo.c]: Done

File: lib/os_cpuinfo.c (hunk 4 above)
Comment (babumoger) [reply lib/os_cpuinfo.c]: Done

**Comment by rkanagar** [comment]
Merged in v24.05 release

**Pull request closed without merging**